broker.go

```diff
@@ -324,11 +326,30 @@ func (b *Broker) cacheMetadata(resp *proto.MetadataResp) {
 			dest := topicPartition{topic.Name, part.ID}
 			b.metadata.endpoints[dest] = part.Leader
 			debugmsg = append(debugmsg, dest, part.Leader)
+
+			count, ok := b.metadata.partitions[topic.Name]
+			if !ok || part.ID >= count {
+				b.metadata.partitions[topic.Name] = part.ID + 1
+			}
 		}
 	}
 	b.conf.Logger.Debug("new metadata cached", debugmsg...)
 }
 
+// GetPartitionCount returns how many partitions a given topic has. If a topic
+// is not known, 0 and an error are returned.
+func (b *Broker) GetPartitionCount(topic string) (int32, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	count, ok := b.metadata.partitions[topic]
+	if ok {
+		return count, nil
+	} else {
+		return 0, errors.New(fmt.Sprintf("topic %s not found in metadata", topic))
```

Comment:
No need for `else`, because `return` is finishing other path.

Use `fmt.Errorf` instead of `errors.New` and `fmt.Sprintf`.

This can return error if topic exists but no metadata was ever fetched. This should check for this case.